src/transforms/key_value_parser.rs

```diff
@@ -0,0 +1,397 @@
+use super::Transform;
+use crate::{
+    config::{DataType, TransformConfig, TransformContext, TransformDescription},
+    event::{self, Event},
+    internal_events::{KeyValueEventProcessed, KeyFailedParse, KeyValueEventFailed},
+    types::{parse_conversion_map, Conversion},
+};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::str;
+use string_cache::DefaultAtom as Atom;
+
+#[derive(Debug, Default, Derivative, Deserialize, Serialize)]
+#[serde(default, deny_unknown_fields)]
+pub struct KeyValueConfig {
+    #[derivative(Default(value = "true"))]
+    pub drop_field: bool,
+    pub field: Option<Atom>,
+    #[derivative(Default(value = "="))]
+    pub field_split: String,
+    #[derivative(Default(value = "true"))]
+    pub overwrite_target: bool,
+    #[derivative(Default(value = " "))]
+    pub separator: String,
+    pub target_field: Option<Atom>,
+    pub trim_key: Option<String>,
+    pub trim_value: Option<String>,
+    pub types: HashMap<Atom, String>,
+}
+
+inventory::submit! {
+    TransformDescription::new::<KeyValueConfig>("key_value_parser")
+}
+
+#[typetag::serde(name = "key_value_parser")]
+impl TransformConfig for KeyValueConfig {
+    fn build(&self, _cx: TransformContext) -> crate::Result<Box<dyn Transform>> {
+        let field = self
+            .field
+            .as_ref()
+            .unwrap_or(&event::log_schema().message_key());
+
+        let conversions = parse_conversion_map(&self.types)?;
+        let trim_key = self.trim_key.as_ref().map(|key| key.chars().collect());
+        let trim_value = self.trim_value.as_ref().map(|key| key.chars().collect());
+
+        Ok(Box::new(KeyValue {
+            conversions,
+            drop_field: self.drop_field,
```

Comment:
I think I missed seeing this before. This should only be true if `field != target_field`:
```rust
        let drop_field = self.drop_field && match &self.target_field {
            Some(target_field) => field != target_field,
            None => true,
        };
```
or
```rust
        let drop_field = self.drop_field && self.target_field.as_ref().map(|target_field| field != target_field).unwrap_or(true);
```
This is to avoid dropping the target that was just parsed.

Reply:
Yes, this is a great check. Thanks!